Refactor `ProductQuery::merge_queries` to accept multiple queries

dinhtungdu opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

<!-- Reference any related issues or PRs here -->

This PR refactors `ProductQuery::merge_queries` method to accept multiple arguments. This makes `merge_queries` the single source of trust in merging queries for Product Query blocks. This PR also improves the readability of the `ProductQuery` class.

More context: https://github.com/woocommerce/woocommerce-blocks/pull/7682#discussion_r1024405565

Fixes #7745

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

Ensure products and order of products are the same as before refactoring.

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [x] Feature plugin
* [ ] Experimental

### Performance Impact
N/a
<!-- Please document any known performance impact (positive or negative) here. If negative, provide some rationale for why this is an okay tradeoff or how this will be addressed. -->

dinhtungdu committed:
try: refactor merge_queries

dinhtungdu committed:
remove debug code

dinhtungdu committed:
Merge branch 'trunk' into refactor/merge_queries

dinhtungdu committed:
Merge branch 'trunk' into refactor/merge_queries

dinhtungdu committed:
Merge branch 'trunk' into refactor/merge_queries

dinhtungdu committed:
fix: unpack assoc array php 7.4

dinhtungdu committed:
try: refactor merge_queries to take any form of input without unpacking
and preparing the input arrays

dinhtungdu committed:
add missing query vars

gigitux commented on src/BlockTypes/ProductQuery.php:
```diff
@@ -186,16 +170,68 @@ function( $acc, $query ) {
+				'suppress_filters',
+				'tax_query',
+			)
+		);
```
Maybe it is premature optimization, but we recreate `$valid_query_vars` array for each query. Isn't it better to move the creation of `$valid_query_vars` outside this function?

dinhtungdu replied:
Thank you for the head-up. I addressed this in [0a3bb43](https://github.com/woocommerce/woocommerce-blocks/pull/7697/commits/0a3bb4347bc71345ab2f3f4b284824f91ed5d903). To me it's more readability than performance improvement as we don't process more than two level depth arrays at the moment, and we still need to recreate the array each time the class is initialized. But it makes the code easier to read so it makes sense to me to implement the optimization.

gigitux reviewed (approved):
Thanks a lot for this refactor! It improved a lot the readability of the code! 💪 

I did an observation, but it isn't a blocker. Feel free to merge it!

dinhtungdu committed:
cache valid query vars

dinhtungdu commented:
@sunyatasattva Can I have your review on this PR as well, as you're working on [Adding support for filtering by attributes within the block](https://github.com/woocommerce/woocommerce-blocks/pull/7743/files#top)?

dinhtungdu committed:
fix merging issue with array_merge_recursive

dinhtungdu committed:
meta query should be nested arrays

dinhtungdu committed:
always merge numeric keys

dinhtungdu committed:
tax_query should contain single nested array too

sunyatasattva commented on src/BlockTypes/ProductQuery.php:
```diff
@@ -186,16 +177,43 @@ function( $acc, $query ) {
+					return $acc;
+				}
+				// If the $query doesn't contain any valid query keys, we unpack it then merge.
+				if ( empty( array_intersect( $this->get_valid_query_vars(), array_keys( $query ) ) ) ) {
+					return $this->merge_queries( $acc, ...array_values( $query ) );
+				}
+				return $this->array_merge_recursive_replace_non_array_properties( $acc, $query );
```
Would you mind explaining a little bit more about this logic?

1. What do you mean we “unpack it”?
2. What are the use-cases for non valid query keys? Are they our custom `__woocommerce` queries?
3. Why do we recursively replace them?

I'd just like to understand the entire flow a bit more and why these things are needed (and perhaps it could be useful to document them in the PHPDocs for posterity).

dinhtungdu replied:
> 1. What do you mean we “unpack it”?

It's restructuring. I used unpack as PHP uses that word. I updated the comment to include both (`unpack/destructure`).

> 3. Why do we recursively replace them?

We recursively replace only non-array values. `array_merge_recursive()`, which I used initially, will merge the value with the same key, but it's not what we want. For example:
```
$base = ['orderby' => 'date']
$new  = ['orderby' => 'meta_value_num']

With array_merge_recursive(), the result will be: ['orderby' => ['date', 'meta_value_num']]

But we want this: ['orderby' => 'meta_value_num']
```

> 2. What are the use-cases for non valid query keys? Are they our custom __woocommerce queries?

_Explaining this makes me think I might be overthinking._

> Are they our custom __woocommerce queries?

No. Query keys (or query vars) here are WP_Query variables. Using valid here is misleading, it should be `built-in` instead. I check if the current array contains the built-in query vars to deal with unknown incoming arrays. For example:

If we feed `merge_queries` with an array like:

```
[
	[
		'meta_query' => [
			...<meta_queries_here>
		]
	]
]
```
`merge_queries` will restructure the array until it gets an array containing the `meta_query` key, which is a built-in query variable, then starts merging.

But, as this is a private method, which is only used inside the ProductQuery class, so we can control the inputs of `merge_queries` calls, I think this logic can be simplified by destructuring until we get the associative arrays, new commit is coming...

dinhtungdu replied:
> But, as this is a private method, which is only used inside the ProductQuery class, so we can control the inputs of `merge_queries` calls, I think this logic can be simplified by destructuring until we get the associative arrays, new commit is coming...

We can't do this because we're returning associative arrays, for example `get_queries_by_attributes()` or `get_queries_by_applied_filters()`.

```
	private function get_queries_by_applied_filters() {
		return array(
			'price_filter'        => $this->get_filter_by_price_query(),
			'attributes_filter'   => $this->get_filter_by_attributes_query(),
			'stock_status_filter' => $this->get_filter_by_stock_status_query(),
		);
	}

```

We can check and unpack/destructure array contains those keys (`price_filter`, `attributes_filter`, `stock_status_filter`), but we will have to update the logic every time we add more queries.

dinhtungdu replied:
After considering, I reverted e6df617. I think the benefit of a versatile `merge_queries()` method outperforms the complexity checking for valid query variables.

sunyatasattva replied:
Thanks a lot @dinhtungdu for the amazing explanation and in-depth commentary! This was extremely helpful to understand this somewhat convoluted logic. And I agree that simplifying is better in this case: glad that some rubberducking was also useful to you.

Just a small note:

>> What do you mean we “unpack it”?

> It's restructuring. I used unpack as PHP uses that word. I updated the comment to include both (unpack/destructure).

By this you mean: `...array_values( $query )` right? In which case, I guess we call this “unpacking” “spreading” and not “destructuring” in the JS world:

> Destructuring is used to create varibles from array items or object properties. Spread syntax is used to unpack iterables such as arrays, objects, and function calls.

Just in case you want to update your comment. But it's not a big deal. Thanks a lot for unpacking ( 🤭 ) the code for me.

dinhtungdu replied:
> By this you mean: `...array_values( $query )` right? In which case, I guess we call this “unpacking” “spreading” and not “destructuring” in the JS world:

> Just in case you want to update your comment. But it's not a big deal. Thanks a lot for unpacking ( 🤭 ) the code for me.

I usually misuse those terms 🤦🏽. I updated the code. Thanks so much, it's easier to memorize for me now 🥰

sunyatasattva commented:
Thanks @dinhtungdu, it looks really good to me. I just have a clarification question, but otherwise the code seems very solid and I like this API more! 👏

dinhtungdu committed:
explain the merge strategy

dinhtungdu committed:
simplify merge logic to unpack/destructure arrays

Instead of checking if array contains built-in query variables, we only
check if the array contains string keys as we can control the inputs of
merge_queries.

dinhtungdu committed:
Revert "simplify merge logic to unpack/destructure arrays"

This reverts commit e6df617b5cf070405e4c87f616d0e94af7562aa7.

gigitux reviewed (approved):
Thanks for providing a great documentation! LGTM!

dinhtungdu committed:
use correct term, thanks Lucio

dinhtungdu committed:
Merge branch 'trunk' into refactor/merge_queries
